tests/tests/classes/PluginManager.js

```diff
@@ -0,0 +1,299 @@
+
+      expect(pluginManager._pluginInstances.length).to.be.at.least(1);
+    });
+  });
+
+  describe('#loadServicePlugins()', () => {
+    it('should load the service plugins', () => {
+      const servicePlugins = [ServicePluginMock1, ServicePluginMock2];
+      pluginManager._loadServicePlugins(servicePlugins);
+
+      expect(pluginManager._pluginInstances.length).to.equal(2);
+    });
+  });
+
+  describe('#loadCommands()', () => {
+    it('should load the plugin commands', () => {
+      pluginManager._loadCommands(SynchronousPluginMock);
+
+      expect(pluginManager._commandsList[0]).to.have.property('deploy');
+    });
+  });
+
+  describe('#getEvents()', () => {
+    beforeEach(() => {
+      pluginManager._loadCommands(SynchronousPluginMock);
+    });
+
+    it('should get all the matching events for a root level command', () => {
+      const commandsArray = 'deploy'.split(' ');
+      const events = pluginManager._getEvents(commandsArray, pluginManager._commands);
+
+      // Note: We expect at least 3 because 3 events will be created for each lifeCycleEvent
+      expect(events.length).to.be.at.least(3);
+    });
+
+    it('should get all the matching events for a nestec level command', () => {
+      const commandsArray = 'deploy onpremises'.split(' ');
+      const events = pluginManager._getEvents(commandsArray, pluginManager._commands);
+
+      // Note: We expect at least 3 because 3 events will be created for each lifeCycleEvent
+      expect(events.length).to.be.at.least(3);
+    });
+
+    it('should return an empty events array when the command is not defined', () => {
+      const commandsArray = 'nonExistingCommand'.split(' ');
+      const events = pluginManager._getEvents(commandsArray, pluginManager._commands);
+
+      expect(events.length).to.equal(0);
+    });
+  });
+
+  describe('#runCommand()', () => {
+    it('should not care about correct capitalization of function name inside a hook', () => {
+      class WrongCapitalizedHookPluginMock {
+        constructor() {
+          this.commands = {
+            deploy: {
+              usage: 'Deploy to the default infrastructure',
+              lifeCycleEvents: [
+                'functions',
+              ],
+            },
+          };
+
+          this.hooks = {
+            // should be "beforeFunctions" but the PluginManager should not bother
+            'deploy:beforefunctions': this.functions.bind(this)
+          };
+
+          // used to test if the function was executed correctly
+          this._deployedFunctions = 0;
+        }
+
+        functions() {
+          this._deployedFunctions = this._deployedFunctions + 1;
+        }
+      }
+
+      pluginManager._addPlugin(WrongCapitalizedHookPluginMock);
+
+      const command = 'deploy';
+      pluginManager.runCommand(command);
+
+      expect(pluginManager._pluginInstances[0]._deployedFunctions).to.equal(1);
+    });
+
+    it('should do nothing when the given command is not available', () => {
```

Comment:
if a user types `serverless random command` and it doesn't exist in any of the plugins, we throw a "command doesn't exist error" ... I think we should keep that pattern otherwise the user won't know what happened.